### PR TITLE
chore: adr in code and junit5 parameterized tests (WPB-10396)

### PR DESCRIPTION
**Pull request by yamilmedina**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10396" title="WPB-10396" target="_blank">StoryWPB-10396</a>  ADR revisited
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

- Add ADRs
- Add junit5 Parameterized tests, and its ADR at `docs/adr/0003-introducing-junit5-parametrizable-tests.md`

### Issues

As part of the discussion in the last collective, this is the revamp ADRs effort, so we can have it in our review process and more important whenever we are including at least these 2 important categories:

- Refactor that changes the way we do something or the software design (see. )
- New library introduced, or replaced.

### Causes (Optional)

- New ADRs needs to be added under `docs/adr` folder.
- The PR will be commented by the action if there is a new ADR document, to bring visibility to the reviewers.
- When merged to develop, this will generate a web page with all the ADRs the project has: https://wireapp.github.io/wire-android

### Attachments (Optional)

https://wireapp.github.io/wire-android/

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by yamilmedina** [comment]
/tmp/new

**Comment by yamilmedina** [comment]
# Decision record template by Michael Nygard

This is the template in [Documenting architecture decisions - Michael Nygard](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions).
You can use [adr-tools](https://github.com/npryce/adr-tools) for managing the ADR files.

In each ADR file, write these sections:

# Title

## Status

What is the status, such as proposed, accepted, rejected, deprecated, superseded, etc.?

## Context

What is the issue that we're seeing that is motivating this decision or change?

## Decision

What is the change that we're proposing and/or doing?

## Consequences

What becomes easier or more difficult to do because of this change?
# 1. Record architecture decisions

Date: 2024-08-05

## Status

Accepted

## Context

We agreed in the past to use ADR's, but we lost track of it as we were using confluence to keep
them. This concern was raised in the last collective, and we need to decide how to proceed.

## Decision

We will use Architecture Decision Records in the code and as part of the review process.
We will use the [Lightway ADR template](0000-template-lightway-adr.md) to keep the ADRs simple and
easy to maintain.

## Consequences

- We need to add a new folder to the repository, `doc/adr`, to keep the architecture decision
  records.
- Whenever a new refactoring or library is introduced, a new ADR should be created.
- You can always request in the Pull request review process to add a new ADR, if you think it's
  necessary.
# 2. Calling activity refactor

Date: 2024-08-01

## Status

Accepted

## Context

To support a second incoming call we need to refactor the code so we can handle the ongoing content
context, without losing the current context.

This is a retroactive decision record implemented
on https://github.com/wireapp/wire-android/pull/3264

## Decision

Create 2 separate activities, one for the Incoming/Outgoing calls and another for the ongoing call.
In this way, we can keep the context of the ongoing call and handle the incoming/outgoing calls.

The design and interaction will look like this:



## Consequences

- StartingActivity will handle Incoming and Outgoing calls content, these contents are disposable
  and can be recreated when receiving a new call.
- OngoingCallActivity will handle the ongoing call content, this content is not disposable and
  should be kept during the call.
# 3. Use parameterizable tests in JUnit5

Date: 2024-08-05

## Status

Proposed

## Context

Sometimes we need to write multiple tests for the same scenario, changing only the input values.

## Decision

We will use parameterizable tests in JUnit5 to avoid writing multiple tests for the same scenario.

## Consequences

- Introduction of `@ParameterizedTest` annotation in the test class
  and [library](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).
- The test method will receive the parameters as arguments.

**Comment by yamilmedina** [comment]
### dddd

**Comment by yamilmedina** [comment]
# Decision record template by Michael Nygard

This is the template in [Documenting architecture decisions - Michael Nygard](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions).
You can use [adr-tools](https://github.com/npryce/adr-tools) for managing the ADR files.

In each ADR file, write these sections:

# Title

## Status

What is the status, such as proposed, accepted, rejected, deprecated, superseded, etc.?

## Context

What is the issue that we're seeing that is motivating this decision or change?

## Decision

What is the change that we're proposing and/or doing?

## Consequences

What becomes easier or more difficult to do because of this change?
# 1. Record architecture decisions

Date: 2024-08-05

## Status

Accepted

## Context

We agreed in the past to use ADR's, but we lost track of it as we were using confluence to keep
them. This concern was raised in the last collective, and we need to decide how to proceed.

## Decision

We will use Architecture Decision Records in the code and as part of the review process.
We will use the [Lightway ADR template](0000-template-lightway-adr.md) to keep the ADRs simple and
easy to maintain.

## Consequences

- We need to add a new folder to the repository, `doc/adr`, to keep the architecture decision
  records.
- Whenever a new refactoring or library is introduced, a new ADR should be created.
- You can always request in the Pull request review process to add a new ADR, if you think it's
  necessary.
# 2. Calling activity refactor

Date: 2024-08-01

## Status

Accepted

## Context

To support a second incoming call we need to refactor the code so we can handle the ongoing content
context, without losing the current context.

This is a retroactive decision record implemented
on https://github.com/wireapp/wire-android/pull/3264

## Decision

Create 2 separate activities, one for the Incoming/Outgoing calls and another for the ongoing call.
In this way, we can keep the context of the ongoing call and handle the incoming/outgoing calls.

The design and interaction will look like this:



## Consequences

- StartingActivity will handle Incoming and Outgoing calls content, these contents are disposable
  and can be recreated when receiving a new call.
- OngoingCallActivity will handle the ongoing call content, this content is not disposable and
  should be kept during the call.
# 3. Use parameterizable tests in JUnit5

Date: 2024-08-05

## Status

Proposed

## Context

Sometimes we need to write multiple tests for the same scenario, changing only the input values.

## Decision

We will use parameterizable tests in JUnit5 to avoid writing multiple tests for the same scenario.

## Consequences

- Introduction of `@ParameterizedTest` annotation in the test class
  and [library](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).
- The test method will receive the parameters as arguments.

**Comment by yamilmedina** [comment]
## ADR Docs in this PR 📚:
# Decision record template by Michael Nygard

This is the template in [Documenting architecture decisions - Michael Nygard](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions).
You can use [adr-tools](https://github.com/npryce/adr-tools) for managing the ADR files.

In each ADR file, write these sections:

# Title

## Status

What is the status, such as proposed, accepted, rejected, deprecated, superseded, etc.?

## Context

What is the issue that we're seeing that is motivating this decision or change?

## Decision

What is the change that we're proposing and/or doing?

## Consequences

What becomes easier or more difficult to do because of this change?
# 1. Record architecture decisions

Date: 2024-08-05

## Status

Accepted

## Context

We agreed in the past to use ADR's, but we lost track of it as we were using confluence to keep
them. This concern was raised in the last collective, and we need to decide how to proceed.

## Decision

We will use Architecture Decision Records in the code and as part of the review process.
We will use the [Lightway ADR template](0000-template-lightway-adr.md) to keep the ADRs simple and
easy to maintain.

## Consequences

- We need to add a new folder to the repository, `doc/adr`, to keep the architecture decision
  records.
- Whenever a new refactoring or library is introduced, a new ADR should be created.
- You can always request in the Pull request review process to add a new ADR, if you think it's
  necessary.
# 2. Calling activity refactor

Date: 2024-08-01

## Status

Accepted

## Context

To support a second incoming call we need to refactor the code so we can handle the ongoing content
context, without losing the current context.

This is a retroactive decision record implemented
on https://github.com/wireapp/wire-android/pull/3264

## Decision

Create 2 separate activities, one for the Incoming/Outgoing calls and another for the ongoing call.
In this way, we can keep the context of the ongoing call and handle the incoming/outgoing calls.

The design and interaction will look like this:



## Consequences

- StartingActivity will handle Incoming and Outgoing calls content, these contents are disposable
  and can be recreated when receiving a new call.
- OngoingCallActivity will handle the ongoing call content, this content is not disposable and
  should be kept during the call.
# 3. Use parameterizable tests in JUnit5

Date: 2024-08-05

## Status

Proposed

## Context

Sometimes we need to write multiple tests for the same scenario, changing only the input values.

## Decision

We will use parameterizable tests in JUnit5 to avoid writing multiple tests for the same scenario.

## Consequences

- Introduction of `@ParameterizedTest` annotation in the test class
  and [library](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).
- The test method will receive the parameters as arguments.

**Comment by AndroidBob** [comment]
# New ADR(s) in this PR 📚:
# Decision record template by Michael Nygard

This is the template in [Documenting architecture decisions - Michael Nygard](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions).
You can use [adr-tools](https://github.com/npryce/adr-tools) for managing the ADR files.

In each ADR file, write these sections:

# Title

## Status

What is the status, such as proposed, accepted, rejected, deprecated, superseded, etc.?

## Context

What is the issue that we're seeing that is motivating this decision or change?

## Decision

What is the change that we're proposing and/or doing?

## Consequences

What becomes easier or more difficult to do because of this change?
# 1. Record architecture decisions

Date: 2024-08-05

## Status

Accepted

## Context

We agreed in the past to use ADR's, but we lost track of it as we were using confluence to keep
them. This concern was raised in the last collective, and we need to decide how to proceed.

## Decision

We will use Architecture Decision Records in the code and as part of the review process.
We will use the [Lightway ADR template](0000-template-lightway-adr.md) to keep the ADRs simple and
easy to maintain.

## Consequences

- We need to add a new folder to the repository, `docs/adr`, to keep the architecture decision
  records.
- Whenever a new refactoring or library is introduced, a new ADR should be created.
- You can always request in the Pull request review process to add a new ADR, if you think it's
  necessary.
# 2. Calling activity refactor

Date: 2024-08-01

## Status

Accepted

## Context

To support a second incoming call we need to refactor the code so we can handle the ongoing content
context, without losing the current context.

This is a retroactive decision record implemented
on https://github.com/wireapp/wire-android/pull/3264

## Decision

Create 2 separate activities, one for the Incoming/Outgoing calls and another for the ongoing call.
In this way, we can keep the context of the ongoing call and handle the incoming/outgoing calls.

The design and interaction will look like this:



## Consequences

- StartingActivity will handle Incoming and Outgoing calls content, these contents are disposable
  and can be recreated when receiving a new call.
- OngoingCallActivity will handle the ongoing call content, this content is not disposable and
  should be kept during the call.
# 3. Use parameterizable tests in JUnit5

Date: 2024-08-05

## Status

Proposed

## Context

Sometimes we need to write multiple tests for the same scenario, changing only the input values.

## Decision

We will use parameterizable tests in JUnit5 to avoid writing multiple tests for the same scenario.

## Consequences

- Introduction of `@ParameterizedTest` annotation in the test class
  and [library](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).
- The test method will receive the parameters as arguments.

File: .github/workflows/deploy-adr-docs.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+name: Deploy ADR Docs
+
+on:
+  push:
+    branches: [ 'develop', 'chore/adr-and-tests' ]
```
Comment (yamilmedina) [inline .github/workflows/deploy-adr-docs.yml:5]: TODO: Remove `'chore/adr-and-tests' ` before merge

File: docs/adr/0003-introducing-junit5-parametrizable-tests.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# 3. Use parameterizable tests in JUnit5
+
+Date: 2024-08-05
+
+## Status
+
+Proposed
```
Comment (yamilmedina) [inline docs/adr/0003-introducing-junit5-parametrizable-tests.md:7]: TODO: Change to Accepted when PR approved

File: app/src/test/kotlin/com/wire/android/ui/userprofile/common/UsernameMapperTest.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    fun `should map other user to its username - handle accordingly`(params: TestParams) {
+        assertEquals(params.expected, UsernameMapper.fromOtherUser(params.input), "Failed for input: <${params.input}>")
+    }
+
+    companion object {
+
+        enum class TestParams(val input: OtherUser, val expected: String) {
+            FEDERATED_USER(OTHER_USER.copy(userType = UserType.FEDERATED, handle = "handle"), "handle@domain"),
+            REGULAR_USER(OTHER_USER.copy(userType = UserType.GUEST, handle = "handle"), "handle"),
+            NO_HANDLE_USER(OTHER_USER.copy(userType = UserType.INTERNAL, handle = null), "")
```
Comment (alexandreferris) [inline app/src/test/kotlin/com/wire/android/ui/userprofile/common/UsernameMapperTest.kt:23]: 😍

**Review by alexandreferris — APPROVED** [review]
Amazing! 😍  looking forward for it

**Review by Garzas — APPROVED** [review]

**Review by vitorhugods — APPROVED** [review]
❤️

**Pull request merged**